### ISSUE TITLE
random spacing issue in output

### ISSUE THREAD

**Issue by plockaby** [dalibo/pg_activity#208]
I'm running pg_activity 2.1.0 from the postgres debian apt repository. It's putting spaces at random on query strings and make it very hard to read what is going on. This only started happening with 2.1.0.

<img width="413" alt="Screen Shot 2021-05-03 at 10 22 04 AM" src="https://user-images.githubusercontent.com/633173/116909852-a4802500-abf9-11eb-8e33-509ebd3b3297.png">

**Comment by dlax** [comment]
> This only started happening with 2.1.0.

Which version were you using before?

I'd need more information about your environment: operating system, terminal emulator, font configuration in the terminal, locale and encoding information.

**Comment by plockaby** [comment]
Ah, very good questions. Sorry for the lack of information. Here's what I have so far.

The previous version that seemed to work OK was 1.4.0-1 from the Debian 10 "buster" archive. The version that I'm running now, that seems to have a problem, is on Debian 10 "buster", all against PostgreSQL 13. I'm using version `2.1.0-1.pgdg100+1` installed from `deb http://apt.postgresql.org/pub/repos/apt/ buster-pgdg main`. I tried a few different configurations and this is what I've come up with. I use iTerm2 as my primary terminal emulator. The font is Monaco Regular 11, anti-aliased. I've also seen it happen on macOS Terminal on both Big Sur and Catalina.

```
TERM=xterm-256color
LANG=en_US.UTF-8
```

I was also seeing it happen in this mode:

```
TERM=xterm-256color
LANG=C.UTF-8
```

Further testing was kind of all over the place. On one server I got it to happen very consistently at a terminal width of 144 but not 143. On another server 144 was a-ok but terminal width 117 caused the display issues.

I manually installed pg_activity 2.0.0 from pip into a virtual environment and it experiences the same issue. When I installed pg_activity 1.4.0 in a virtual environment the problem did not appear.

I hope that provides some more context. Let me know if there is anything else that I can provide.

**Comment by dlax** [comment]
If I'm reading correctly, you're using a terminal on macOS but pg_activity is running on a Debian system, so I suppose you're using an SSH connection, right?

It'd then be helpful to try either directly on the Debian server or use an SSH connection from another (preferably Linux) system.

For the record, pg_activity was almost completely rewritten in version 2.0, so quite a few things have changed (noticeably, we're now using a third-party library, [blessed](https://blessed.readthedocs.io/), for terminal rendering).

**Comment by plockaby** [comment]
I launched a virtual machine running Debian bullseye in VMWare Fusion and then from that virtual machine's console (not via SSH) I connected to my database server over SSH and ran pg_activity directly on the database server, as usual, and the display looks fine. Of course, being a console I am not scale it to try different screen sizes. I'm not sure if that helps, though. I suppose it's worth noting that only the query includes spaces. No other fields have spacing issues.

**Comment by dlax** [comment]
> I suppose it's worth noting that only the query includes spaces. No other fields have spacing issues.

Ok, let's start from that.
Looking back at your screenshot, it may be that your relation name ("ma sk_ident ...") contains Unicode whitespace characters. (We currently escape those by a "space", and I'm not sure it's still useful.)
Can you check that?

**Comment by plockaby** [comment]
Ah I'm very sure that it is not what is happening. In this query right here you can see that it is happening even in keywords. 
<img width="1025" alt="Screen Shot 2021-05-10 at 6 00 14 PM" src="https://user-images.githubusercontent.com/633173/117742501-c9583780-b1b9-11eb-8d85-b216f49d06f4.png">

**Comment by plockaby** [comment]
Excellent. I cannot replicate the problem with ```--verbose-mode=1``` enabled no matter what width I set my terminal to.

**Comment by dlax** [comment]
Thanks; now I'm pretty sure the issue is about blessed, the terminal rendering library we're using, especially the [text wrapper]( https://blessed.readthedocs.io/en/latest/api/sequences.html#blessed.sequences.SequenceTextWrapper).

**Comment by dlax** [comment]
Commit 81aa0b850b9c912c08f116aa5d96cfeea2f2ec35 contains a possible fix.

@plockaby, before we ship it, can you try it out? You can get it, in a (temporary) virtualenv, as:

    pip install https://github.com/dalibo/pg_activity/archive/refs/heads/query-wrap.zip

Then use pg_activity as normal, without `--verbose-mode=1` especially.
Thanks.

**Comment by plockaby** [comment]
That's great. I cannot reproduce the problem with that version.

**Comment by plockaby** [comment]
Sorry, I take that back. I did just make it happen with the new version.

**Comment by plockaby** [comment]
Sorry to keep replying but I figured that I would mention that spaces that are appearing are appearing in different places in the two versions. The first of these images is from the latest version on pip and the second is from the version you just sent me.
![From the old version.](https://user-images.githubusercontent.com/633173/118009397-df254400-b302-11eb-8ad3-3c4b7511f6e6.png)
![From the new version.](https://user-images.githubusercontent.com/633173/118009352-d92f6300-b302-11eb-8a11-903dc9ccffb5.png)

I'm trying to provide as much of the queries as possible without revealing exactly what queries the users are running so obviously there are some things blacked out in the query.

**Comment by dlax** [comment]
I made another change, can you try back with the same instructions as before (possibly from a fresh virtualenv)?
(If that does not fix the issue, I'll back at this next week.)

**Comment by plockaby** [comment]
That didn't work. I figured that I should stop letting you shoot at solutions completely blind and start digging in myself. I took your latest version and started debugging inline.

This is the exact block that is getting called that is causing problems:

```
                        # Only wrap subsequent lines.
                        wrapped_lines = TextWrapper(dif).wrap(query)
                        if wrapped_lines:
                            query_lines = [wrapped_lines[0]] + TextWrapper(width).wrap(
                                " ".join(wrapped_lines[1:])
                            )
                        else:
                            query_lines = []
```

The query that it is coming out of pg_stat_activity is this, including the spaces:

```
WITH seed_posts as (
    SELECT id
    FROM governor_replies_experiment
    WHERE depth = 1
    AND user_screen_name = 'XXXXXX')

    SELECT date_trunc('day', governor_replies_experiment.created_at) as "day", count(*) as 
    total_mentions
    FROM governor_replies_experiment, seed_posts
    WHERE seed_posts.id = governor_replies_experiment.seed_id
    GROUP BY 1
    ORDER BY 1
```

And then the TextWrapper is getting this as its arguments:
```
dif=10, query=WITH seed_posts as ( SELECT id FROM governor_replies_experiment WHERE depth = 1 AND user_screen_name = 'XXXXXX') SELECT date_trunc('day', governor_replies.created_at) as "day", count(*) as total_mentions FROM governor_replies_experiment, seed_posts WHERE seed_posts.id = governor_replies_experiment.seed_id GROUP BY 1 ORDER BY 1
```

The first call to TextWrapper is generating this:
```
wrapped_lines = ['WITH', 'seed_posts', 'as (', 'SELECT id', 'FROM gover', 'nor_replie', 's_experime', 'nt WHERE', 'depth = 1', 'AND user_s', 'creen_name', "= 'XXX", "XXX')", 'SELECT dat', "e_trunc('d", "ay', gover", 'nor_replie', 's.created_', 'at) as', '"day",', 'count(*)', 'as total_m', 'entions', 'FROM gover', 'nor_replie', 's_experime', 'nt,', 'seed_posts', 'WHERE seed', '_posts.id', '= governor', '_replies_e', 'xperiment.', 'seed_id', 'GROUP BY 1', 'ORDER BY 1']
```

And all of my extra spaces are where those breaks are in the array. I'm going to keep looking to figure out how to rewrite this so that it works correctly.

**Comment by plockaby** [comment]
Ok, this change solves my problem:

```
                        # Only wrap subsequent lines.
                        wrapped_lines = TextWrapper(dif, drop_whitespace=False).wrap(query)
                        if wrapped_lines:
                            query_lines = [wrapped_lines[0]] + TextWrapper(width, drop_whitespace=False).wrap(
                                "".join(wrapped_lines[1:])
                            )
                        else:
                            query_lines = []
```

That is, I added ```drop_whitespace=False``` and changed ```" ".join``` to ```"".join```. Problem solved. :)

**Comment by dlax** [comment]
Ah, nice catch!
If you want to open a pull request with that change, that'd be great. I'd suggest to use the query-wrap branch as a base.
Thanks

**Comment by dlax** [comment]
Fixed by e44cab8ba14b3bb664ae1133f8c915e1027fcb01.
